docs/source/api/apollo-server.md

```diff
@@ -345,10 +345,26 @@ addMockFunctionsToSchema({
 
 *  `privateVariables`: Array<String> | boolean
 
+   DEPRECATING IN VERSION XX.XX.XX for `enforcePrivateVariables`, which will support the same
+   functionalities but allow for more flexibility.
+   
    A case-sensitive list of names of variables whose values should not be sent
    to Apollo servers, or 'true' to leave out all variables. In the former
    case, the report will indicate that each private variable was redacted in
    the latter case, no variables are sent at all.
+   
+* `enforcePrivateVariables`: ((variables: Record<string, any>, operationString?: string) => Record<string, any>) | Array<String> | boolean
```

Comment:
I'm not sure that the term "enforce" makes sense to me here.  That makes it seems like there's some externally defined rules that this is merely the enforcement of, rather than this being the configuration?  ie, if I read `enforcePrivateVariables: true` I would think it means "yes, we should enforce that private variables stay private" or something, not "treat all variables as private".

Honestly I think "private" isn't really the right term here. I think this ought to have a verb rather than an adjective. Like "mask" or "hide" or "redact". `maskVariables`?

On the other hand, just keeping this as `privateVariables` seems fine — is there a strong reason we can't just add alternate possible types to `privateVariables` (and change the default)?

On the other other hand, if the default is "don't send variables", maybe we should choose a name that describes what the option *does* (sends values) not what it doesn't do.  So what if the name was `sendVariableValues`. `true` means send all variable values.  A list means send just the values of these variables.  A function lets you arbitrarily tweak the map. The downside to this is it isn't obvious how to express "all except these values" (ie, the current feature).  We could allow `sendVariableValues: {not: ["foo, "bar"]}` or something for that case?  @zionts 's idea that this always involves a sub-key is nice too.

I also like including "Values" in the name since that makes it clearer that in the default case we still include variable names!

(FWIW I think setting both the old and new options could be a hard error.)

Curious what others think, but I'm definitely leaning towards the idea that the name should say what action happens when you set it ("send variable values"), not an abstract definition of the concept of privacy.

Reply:
Also maybe the custom function should take an options object instead of two arguments, for cleaner extensibility later.

Reply:
I like your idea for `maskVariables` (or maybe  `maskVariableValues`) the best! I thought "enforce" would be alright because we're allowing the user to enforce some custom privacy policy, but you're right, I think it could be clearer.

I also agree that having a verb in the term name feels more appropriate, but I was thinking because the option accepts functions that will make modifications to the variable values. This is one reason why I was worried about continuing with `privateVariables` - I don't think it captures this new functionality well? And the other reason was because of the minor changes to the expected behavior when set to `true` (I wrote more in a response to one of your comments below). 

I think `sendVariableValues` could also work, but I'm worried that because it's sort of the opposite of 'privateVariables`, that people might get confused (e.g. the old defaults to `true`, this one should default to `false`). 

I'll make changes to the arguments, for both the option and for the custom function like you and @zionts suggested!

Reply:
In general I do find that it's kinda awkward to have options whose default is "true" and it's nicer if the default can be false/undefined, which is why I like the idea of `sendVariableValues` being the option rather than `maskVariableValues`. Curious what @zionts thinks.

Reply:
I definitely agree with the idea of choosing a name that's descriptive of the overall action being performed, and I think we could have an obvious-when-read solution for making the easy route (boolean / list options) by using descriptive key names. I think we could have something like 
```
variableMaskingPolicy: {
  redactAllValues: boolean;
} | 
{
  doNothing: boolean;
} |
{
  variableNamesToRedactValues: Array<String>
} |
{
  redactAllKeysAndValues: boolean;
} |
{
  customFunction: (operation: DocumentNode, schema: GraphQLSchema, variables: Record<String, Any>) => Record<String, Any>
}
```

which might help to accomplish both goals? I also think that something David mentioned earlier is that we should default to saving the keys but redacting the values, and I think that's definitely something we **don't** do now, but we should! Similarly, I'm really thinking that we should enforce that the keys that are sent are either always maintained with the operation (to make it, y'know, valid, or at least only **filtered** (i.e. you can't replace the $foo key with $squiggle just because you feel like it), but I can be convinced otherwise.

Reply:
RE descriptive key names: I've put three in the latest commit (variableModifier, exceptVariableNames, safelistAll), but can add more if you think that'll be helpful for clarity!

RE keys sent are maintained with the operation: I added a check to remove any keys that weren't originally in the operation, and to add back (but with no value) keys that were!